lms/services/canvas.py

```diff
@@ -1,41 +1,198 @@
-from lms.services.exceptions import CanvasFileNotFoundInCourse
+from functools import cached_property, lru_cache
+from logging import getLogger
+
+from sqlalchemy import and_
+
+from lms.models import File
+from lms.services.exceptions import CanvasAPIPermissionError, CanvasFileNotFoundInCourse
+
+LOG = getLogger(__name__)
 
 
 class CanvasService:
     """A high level Canvas service."""
 
     api = None
 
-    def __init__(self, canvas_api):
+    def __init__(
+        self, canvas_api, application_instance_service, assignment_service, db_session
+    ):
         self.api = canvas_api
 
-    def public_url_for_file(self, file_id, course_id, check_in_course=False):
+        self._application_instance_service = application_instance_service
+        self._assignment_service = assignment_service
+        self._db_session = db_session
+
+    def public_url_for_file(
+        self, file_id, course_id, resource_link_id, check_in_course=False
+    ):
         """
         Get a public URL for a Canvas file.
 
-        This will also attempt to check if the file is in the course to detect
-        course copy situations and fix it if we can.
+        This will also attempt to use any file with the same size and name if
+        one can be found.
 
         :param file_id: The file to look up
         :param course_id: The course the file should be in
-        :param check_in_course: Raise CanvasFileNotFoundInCourse if file_id isn't in
-            course_id
+        :param resource_link_id: The id of the assignment
+        :param check_in_course: Attempt to map detect and map the file if it's
+            not in the course before attempting to retrieve it. This is more
+            thorough, but causes an extra API request
         :return: A URL suitable for public presentation of the file
-        :raise  CanvasFileNotFoundInCourse: if check_in_course=True and file_id isn't in course_id
-            be in the course but isn't.
+        :raise CanvasFileNotFoundInCourse: if the file is not accessible and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIResourceNotFound: if the file is not valid at all, and
+            cannot be mapped to a file in the course
+        :raise CanvasAPIPermissionError: If we cannot retrieve the file for
+            any other reason
         """
+        # Check to see if we've mapped this file id to something else before
+        file_mapping = self._file_mapping(resource_link_id)
```

Comment:
The file mapping is attached to the assignment, which means we now have to pass through the `resource_link_id` which is the id of the assignment. This is the cause of having to thread it through everything else.

The mapping object here is a small shim around interacting with a specific dict inside a key in the extra object below. It's made to look like a dict in the ways we use it.